hypershift-operator/controllers/nodepool/capi.go

```diff
@@ -0,0 +1,1339 @@
+package nodepool
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	agentv1 "github.com/openshift/cluster-api-provider-agent/api/v1beta1"
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/nodepool/kubevirt"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/nodepool/openstack"
+	"github.com/openshift/hypershift/support/api"
+	supportutil "github.com/openshift/hypershift/support/util"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/validation"
+	k8sutilspointer "k8s.io/utils/pointer"
+	capiaws "sigs.k8s.io/cluster-api-provider-aws/v2/api/v1beta2"
+	capiazure "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	capipowervs "sigs.k8s.io/cluster-api-provider-ibmcloud/api/v1beta2"
+	capikubevirt "sigs.k8s.io/cluster-api-provider-kubevirt/api/v1alpha1"
+	capiopenstack "sigs.k8s.io/cluster-api-provider-openstack/api/v1beta1"
+	capiv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// CAPI Knows how to reconcile all the CAPI resources for a unique token.
+type CAPI struct {
+	*Token
+	capiClusterName string
+}
+
+func newCAPI(token *Token, capiClusterName string) (*CAPI, error) {
+	if token == nil {
+		return nil, fmt.Errorf("token can not be nil")
+	}
+
+	if capiClusterName == "" {
+		return nil, fmt.Errorf("capiClusterName can not be empty")
+	}
+
+	return &CAPI{
+		Token:           token,
+		capiClusterName: capiClusterName,
+	}, nil
+}
+
+func (c *CAPI) Reconcile(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	nodePool := c.nodePool
+	if err := c.cleanupMachineTemplates(ctx, log, nodePool, c.controlplaneNamespace); err != nil {
+		return err
+	}
+
+	//  Reconcile (Platform)MachineTemplate.
+	template, mutateTemplate, _, err := c.machineTemplateBuilders()
+	if err != nil {
+		return err
+	}
+	if result, err := c.CreateOrUpdate(ctx, c.Client, template, func() error {
```

Comment:
I think we are introducing a tight coupling with the `Token` struct by using its client, createOrUpdate, nodePool, etc. fields.
should we explicitly pass those options to the `Reconcile()` functions or add them as fields to `CAPI`?

I would imagine the `Token` object we pass to CAPI to just be an interface exposing the functions CAPI needs.

Reply:
I agree, I'd expect that's exactly the end state (fields in capi + interface) and even consider separate packages so unexported fields are not accessible. I chose coupling for simplicity for now, the fields are accessible within the same package anyways and otherwise you would endup with "duplicated" fields which is also confusing. Any follow up refinement in that direction can be done now with confidence since we have decent coverage.
Added a TODO